lib/oelite/profiling.py

```diff
@@ -0,0 +1,188 @@
+import datetime
+import functools
+import atexit
+import oelite.util
+import inspect
+import os
+import sys
+from resource import *
+
+now = datetime.datetime.utcnow
+
+profiledir = None
+
+# Decorating any function with @profile_calls will record the duration
+# of every call of that function. Some statistics on these are
+# automatically printed to $profiledir/callstats.txt on exit.
+profiled_functions = dict()
+def profile_calls(somefunc):
+    @functools.wraps(somefunc)
+    def recordtime(*args, **kwargs):
+        start = now()
+        try:
+            return somefunc(*args, **kwargs)
+        finally:
+            delta = (now()-start).total_seconds()
+            if not somefunc in profiled_functions:
+                profiled_functions[somefunc] = SimpleStats()
+            profiled_functions[somefunc].append(delta)
+    return recordtime
+
+def write_call_stats():
+    outfn = os.path.join(profiledir, "call_stats.txt")
+    with open(outfn, "w") as out:
+        for f in sorted(profiled_functions.keys(), key=lambda x: x.__name__):
+            name = f.__name__
+            try:
+                srcfile = os.path.basename(inspect.getsourcefile(f))
+            except TypeError:
+                srcfile = "<unknown>"
+
+            stats = profiled_functions[f]
+            stats.compute()
+            out.write("%-12s\t%-24s\t%9.3fs / %5d = %7.3f " %
+                    (srcfile, f.__name__,
+                     stats.sum, stats.count, stats.mean))
+            out.write("[%s]\n" % ", ".join(["%7.3f" % x for x in stats.quartiles]))
```

Comment:
jeeez, please use placeholders instead of doing list comprehension as a function arguement...